src/emqx_auth_ldap_cli.erl

```diff
@@ -63,24 +65,74 @@ connect(Opts) ->
             catch
                 error:Reason -> {error, Reason}
             end;
-        {error, Reason} ->
+        {_, {error, Reason}} ->
             {error, Reason}
     end.
 
 search(Base, Filter) ->
-    ecpool:with_client(?APP, fun(C) ->
-                                 eldap2:search(C, [{base, Base},
-                                                   {filter, Filter},
-                                                   {deref, eldap2:derefFindingBaseObj()}])
-                             end).
+    ecpool:with_client(?APP,
+        fun(C) ->
+                case application:get_env(?APP, bind_as_user) of
+                    {ok, true} ->
+                        {ok, Opts} = application:get_env(?APP, ldap),
+                        BindDn       = get_value(bind_dn, Opts),
+                        BindPassword = get_value(bind_password, Opts),
+                        try eldap2:simple_bind(C, BindDn, BindPassword) of
+                            ok ->
+                                eldap2:search(C, [{base, Base},
+                                                  {filter, Filter},
+                                                  {deref, eldap2:derefFindingBaseObj()}]);
+                            {error, Error} ->
+                                {error, Error}
+                        catch
+                            error:Reason -> {error, Reason}
+                        end;
+                    {ok, false} ->
+                        eldap2:search(C, [{base, Base},
+                                          {filter, Filter},
+                                          {deref, eldap2:derefFindingBaseObj()}])
+                end
+        end).
 
 search(Base, Filter, Attributes) ->
-    ecpool:with_client(?APP, fun(C) ->
-                                 eldap2:search(C, [{base, Base},
-                                                   {filter, Filter},
-                                                   {attributes, Attributes},
-                                                   {deref, eldap2:derefFindingBaseObj()}])
-                             end).
+    ecpool:with_client(?APP,
+        fun(C) ->
+                case application:get_env(?APP, bind_as_user) of
```

Comment:
I think here these two `search/2,3` could be merged into one, if there is no significant difference?  I am not sure if there is an acceptable default value for `attributes`.

Reply:
They are used in different circumstances. `search/3` is used for ACL's, the other is for users. Keeping the original design for now, as it is not covered by tests at all.